Add initial support for Groups and GroupStates

haudren opened the pull request:
This commit adds a new binding for srdf_parser.h
Current support is:
- Retrieving Groups and GroupStates
- Accessing their members
Note that due to SWIG's lack of support of nested classes, they end up
flattened, at the root of the module.

This PR still lacks a few things, namely tests but I'd like to get a few comments on this code first.

Levi-Armstrong commented:
@johnwason Do you mind taking a look at this PR?

Levi-Armstrong commented:
Would moving the structs outside the class help?

Levi-Armstrong commented:
@haudren  Just a side note. You can also get manipulator group names using the `getAvailableFwdKinematicsManipulators` in the forward and inverse kinematics managers.

johnwason commented:
Huh, this didn't work for me when I tried it. Can you read the values in Group and GroupState?

haudren commented:
Interesting, I managed to get the values of the GroupState for sure... I'll have another look.

johnwason commented:
Sometimes SWIG will return an opaque pointer if it doesn't understand the type. Can you read the fields in Group and GroupState?

haudren commented:
OK I just pushed another update, I was indeed missing a couple of template bindings, which led to opaque pointers appearing. I fixed those, and added some tests which show how you can recover the SRDF information from Python.

johnwason commented:
Most of the std templates are already defined in other *.i files. Can you remove the duplicates?

haudren commented:
I would, but how do I go about it? When I did not include them, I believe I got some opaque pointers... I'll give it another go.

Levi-Armstrong commented:
@johnwason What is the best way to address your comment?

johnwason commented:
My suggested corrections:

* Move the std::pair<std::string,std::string> templates to tesseract_python.i where vector_string is defined.
* std::vector<double> has a custom typedef that should be used instead of overriding the definition. See https://github.com/ros-industrial-consortium/tesseract/blob/772ef43cbe4c270deea223560c02d9f35309e4a2/tesseract_python/tesseract_python/swig/trajopt/problem_description.i#L25

The rest look fine.

haudren commented:
So I tried to address your comment @johnwason but I keep on hitting this error
```
error: ‘type_name’ is not a member of ‘swig::traits<double>’
     return traits<typename noconst_traits<Type >::noconst_type >::type_name();
```
Which goes away when I explicitly instantiate the `std::vector<double>` template. I tried to play around with the typemaps:
- Move them from `problem_description.i` to `tesseract_python.i`
- Create my own custom typemap similar to what you did but for `std::vector<double>` (not a pointer)
- Do the above and introduce a typedef to alias to a non-templated type
All to no avail

I think I could use some help here...

Levi-Armstrong commented:
@haudren It looks like CI is passing. Is the issue resolved or are you still waiting on a response from @johnwason?

johnwason commented:
I will take a look at this today.

johnwason commented:
Everything looks good to me. The only problem I found is that the `test_srdf_parser.py` doesn't actually call the test function. The following is required at the bottom of the file:

    if __name__ == "__main__":
        test_environment()

johnwason commented:
CMake sometimes won't detect that a swig .i file has changed if the root file isn't modified. Try touching the tesseract_python.i file and see if that helps.

    touch -c tesseract_python.i

Levi-Armstrong commented:
@johnwason Thank you for looking into this.

Levi-Armstrong commented:
@haudren Looks just one thing to add before this can be merged.

haudren commented:
Fixed it thanks :+1:
